galley/pkg/crd/validation/config.go

```diff
@@ -27,55 +27,103 @@ import (
+		log.Warnf("%s/%v endpoint not ready: no subsets", namespace, name)
+		return false, false
 	}
-
 	for _, subset := range endpoints.Subsets {
 		if len(subset.Addresses) > 0 {
-			return nil
+			return false, true // endpoint ready at least once
 		}
 	}
-	return fmt.Errorf("%s/%v endpoint not ready: no ready addresses", wh.deploymentNamespace, wh.deploymentName)
+	log.Warnf("%s/%v endpoint not ready: no ready addresses", namespace, name)
+	return false, false
 }
 
-func (wh *Webhook) createOrUpdateWebhookConfig() {
-	if wh.webhookConfiguration == nil {
-		log.Error("validatingwebhookconfiguration update failed: no configuration loaded")
-		reportValidationConfigUpdateError(errors.New("no configuration loaded"))
-		return
-	}
-
+func (wh *Webhook) waitForEndpointReady(stop <-chan struct{}) (shutdown bool) {
 	// During initial Istio installation its possible for custom
 	// resources to be created concurrently with galley startup. This
 	// can lead to validation failures with "no endpoints available"
 	// if the webhook is registered before the endpoint is visible to
-	// the rest of the system. Minimize this problem by waiting the
-	// galley endpoint is available at least once before
+	// the rest of the system. Minimize this problem by waiting for the
+	// galley endpoint to be available at least once before
 	// self-registering. Subsequent Istio upgrades rely on deployment
 	// rolling updates to set maxUnavailable to zero.
-	if !wh.endpointReadyOnce {
-		if err := wh.endpointReady(); err != nil {
-			log.Warnf("%v validatingwebhookconfiguration update deferred: %v",
-				wh.webhookConfiguration.Name, err)
-			reportValidationConfigUpdateError(errors.New("endpoint not ready"))
-			return
+	controllerStopC := make(chan struct{})
+	defer close(controllerStopC)
+
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	// TODO - this assumes the deployment and service name match
+	store, controller := cache.NewInformer(
+		wh.createInformerEndpointSource(wh.clientset, wh.deploymentNamespace, wh.deploymentName),
+		&v1.Endpoints{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+		},
+	)
+
+	go controller.Run(controllerStopC)
+
+	if !cache.WaitForCacheSync(stop, controller.HasSynced) {
+		return true
+	}
+
+	for {
+		shutdown, ready := endpointReady(store, queue, wh.deploymentNamespace, wh.deploymentName)
+		if shutdown {
+			return true
+		} else if ready {
+			return false
+		} else {
+			// continue
```

Comment:
```
                if shutdown || ready{
			return
		} 
```

Reply:
I'm not sure how to interpret this comment - typo?

Reply:
a more readable way